### DISCUSSION TITLE
Improving DotLottie error handling

### DISCUSSION THREAD

**Discussion by mkj-is** [airbnb/lottie-ios#1957] [Ideas]
We faced some issues related to failing animation loads using async DotLottie methods (#1956).

I figured error handling in DotLottie loading can be improved in some ways:

- Using `LottieLogger` to report any errors which are silent.
- Return internal error as part of `handleResult` (`DotLottieError.invalidData` is too generic for debugging purposes and will be returned even in case there was some issue with opening/decompressing `.lottie` file or manifest).

Places where error reporting can be improved:

- https://github.com/airbnb/lottie-ios/blob/master/Sources/Public/DotLottie/DotLottieFile.swift#L45
- https://github.com/airbnb/lottie-ios/blob/master/Sources/Public/DotLottie/DotLottieFile.swift#L128
- https://github.com/airbnb/lottie-ios/blob/master/Sources/Public/DotLottie/DotLottieFileHelpers.swift#L245

Do you think it would be better to send errors to LottieLogger or return them in `handleResult` closure?

I am willing to help with the implementation.

**Comment by calda** [comment]
I think it would make sense to log these errors as either `LottieLoggier.shared.assertionFailure` or `LottieLogger.shared.warn`.

**Reply by calda** [reply]
It would also be nice to add context about the underlying error to the various `DotLottieError` cases.

**Reply by mkj-is** [reply]
Great, I will open a draft PR on which we can iterate.

**Reply by mkj-is** [reply]
Opened a PR: https://github.com/airbnb/lottie-ios/pull/1969

I am not wrapping the underlying errors since public loading methods return the generic `Error`.

It actually made the code shorter and I was able to get rid of some unused helper variables.
